File: 백준/Silver/1012. 유기농 배추/유기농 배추.cc
#include <iostream>
#include <vector>
#include <queue>

using namespace std;

int T, M, N, K;
vector<vector<int>> maps;
vector<vector<int>> visited;
vector<pair<int, int>> direction = {{-1,0}, {0,-1}, {1,0}, {0,1}};

void bfs(int row, int col) {
    queue<pair<int, int>> q;
    q.push({row, col});
    visited[row][col] = 1;

    while (!q.empty()) {
        auto [r, c] = q.front(); q.pop();
        for (auto [dr, dc] : direction) {
            int nr = r + dr;
            int nc = c + dc;
            if (0 <= nr && nr < N && 0 <= nc && nc < M) {
                if (maps[nr][nc] == 1 && visited[nr][nc] == 0) {
                    visited[nr][nc] = 1;
                    q.push({nr, nc});
                }
            }
        }
    }
}

int main() {
    cin >> T;

    while (T--) {
        cin >> M >> N >> K;
        maps.assign(N, vector<int>(M, 0));
        visited.assign(N, vector<int>(M, 0));

        for (int i = 0; i < K; ++i) {
            int x, y;
            cin >> x >> y;
            maps[y][x] = 1;
        }

        int cnt = 0;
        for (int i = 0; i < N; ++i) {
            for (int j = 0; j < M; ++j) {
                if (maps[i][j] == 1 && visited[i][j] == 0) {
                    bfs(i, j);
                    cnt++;
                }
            }
        }

        cout << cnt << '\n';
    }

    return 0;
}

File: 백준/Silver/1012. 유기농 배추/유기농 배추.py
import sys
sys.setrecursionlimit(10**6)

T = int(input())

def dfs(row, col):
    for a, b in direction:
        ncol, nrow = col + a, row + b
        if (0 <= ncol < M) and (0 <= nrow < N) and (maps[nrow][ncol] == 1) and (visited[nrow][ncol] == 0):
            visited[nrow][ncol] = 1
            dfs(nrow, ncol)

for _ in range(T):
    M, N, K = map(int, input().split())
    maps = [[0]*M for _ in range(N)]
    visited = [[0]*M for _ in range(N)]
    direction = [(-1,0), (0,-1), (1,0), (0,1)]

    for _ in range(K):
        X, Y = map(int, input().split())
        maps[Y][X] = 1

    cnt = 0
    for i in range(N):
        for j in range(M):
            if visited[i][j] == 0 and maps[i][j] == 1:
                visited[i][j] = 1
                dfs(i, j)
                cnt += 1

    print(cnt)
